mlflow/models/utils.py

```diff
@@ -1247,26 +1315,29 @@ def _enforce_pyspark_dataframe_schema(
     return new_pf_input.drop(*columns_to_drop)
 
 
-def _enforce_datatype(data: Any, dtype: DataType, required=True):
+def _enforce_datatype(data: Any, dtype: DataType, required=True, enforce_param=False):
     if not required and _is_none_or_nan(data):
         return None
 
     if not isinstance(dtype, DataType):
         raise MlflowException(f"Expected dtype to be DataType, got {type(dtype).__name__}")
     if not np.isscalar(data):
         raise MlflowException(f"Expected data to be scalar, got {type(data).__name__}")
-    # Reuse logic in _enforce_mlflow_datatype for type conversion
-    pd_series = pd.Series(data)
-    try:
-        pd_series = _enforce_mlflow_datatype("", pd_series, dtype)
-    except MlflowException:
-        raise MlflowException(
-            f"Failed to enforce schema of data `{data}` with dtype `{dtype.name}`"
-        )
-    return pd_series[0]
+    if enforce_param:
+        return _enforce_param_datatype(data, dtype)
+    else:
+        # Reuse logic in _enforce_mlflow_datatype for type conversion
+        pd_series = pd.Series(data)
```

Comment:
Basic question: Why we need separate type enforcement logic between param and input data?

Reply:
Because params are mostly used for genai flavors, and they only accept pure python types. While the _enforce_mlflow_datatype method was built based on some numpy data assumption (difference between int32 and int64), and applying this on params will cause error when passing to the model

Reply:
I'm thinking if we should disallow any kind of data conversion in params though (in the future)

Reply:
> While the _enforce_mlflow_datatype method was built based on some numpy data assumption (difference between int32 and int64), and applying this on params will cause error when passing to the model

Interesting, what kind of error does it produce? It sounds strange that we always pass value as single scaler so not sure where that discrepancy happens🤔

Reply:
For example transformers model might not accept numpy.int64 value as int:
<img width="1160" alt="image" src="https://github.com/user-attachments/assets/757597b9-8b45-4771-b2bd-78f151e6ef34" />
_enforce_mlflow_datatype always converts the value to numpy values, even if we only pass a single scalar value there are differences

Reply:
> _enforce_mlflow_datatype always converts the value to numpy values, even if we only pass a single scalar value there are differences

What.... ok thanks for clarification tho. But taking a step back, do we want to 'enforce' parameter type then? For example, can we just use `_enforce_mlflow_datatype` for validation only, and just use the original param values as they are?

Reply:
Don't want to block too much here... but from my observation what users want for params is "please make something works as normally", not a magical type handling😔

Reply:
Yes I agree, I think by the time params was introduced I just followed similar pattern as data to allow some conversion, but all over the end I think params should always be python built-in types. 
I agree on your points, previously I thought we need to update the behavior for non-object params as well (no data conversion), but yea we can make no data conversion on objects in this PR first, and probably drop data conversion for scalar params in MLflow 3.0

Reply:
Yea totally agreed on the plan towards MLflow 3.0. Thanks for the update!